src/core/common/tlvs.cpp

```diff
@@ -38,98 +38,120 @@
 {
-    otError  error  = OT_ERROR_NOT_FOUND;
-    uint16_t offset = aMessage.GetOffset();
-    uint16_t end    = aMessage.GetLength();
-    Tlv      tlv;
-
-    while (offset + sizeof(tlv) <= end)
-    {
-        uint32_t length = sizeof(tlv);
-
-        aMessage.Read(offset, sizeof(tlv), &tlv);
-
-        if (tlv.GetLength() != kExtendedLength)
-        {
-            length += tlv.GetLength();
-        }
-        else
-        {
-            uint16_t extLength;
-
-            VerifyOrExit(sizeof(extLength) == aMessage.Read(offset + sizeof(tlv), sizeof(extLength), &extLength));
-            length += sizeof(extLength) + HostSwap16(extLength);
-        }
+    return Find(aMessage, aType, &aOffset, NULL, NULL);
+}
 
-        VerifyOrExit(offset + length <= end);
+otError Tlv::GetValueOffset(const Message &aMessage, uint8_t aType, uint16_t &aValueOffset, uint16_t &aLength)
+{
+    otError  error;
+    uint16_t offset;
+    uint16_t size;
+    bool     isExtendedTlv;
 
-        if (tlv.GetType() == aType)
-        {
-            aOffset = offset;
-            ExitNow(error = OT_ERROR_NONE);
-        }
+    SuccessOrExit(error = Find(aMessage, aType, &offset, &size, &isExtendedTlv));
 
-        offset += static_cast<uint16_t>(length);
+    if (!isExtendedTlv)
+    {
+        aValueOffset = offset + sizeof(Tlv);
+        aLength      = size - sizeof(Tlv);
+    }
+    else
+    {
+        aValueOffset = offset + sizeof(ExtendedTlv);
+        aLength      = size - sizeof(ExtendedTlv);
     }
 
 exit:
     return error;
 }
 
-otError Tlv::GetValueOffset(const Message &aMessage, uint8_t aType, uint16_t &aOffset, uint16_t &aLength)
+otError Tlv::Find(const Message &aMessage, uint8_t aType, uint16_t *aOffset, uint16_t *aSize, bool *aIsExtendedTlv)
 {
     otError  error  = OT_ERROR_NOT_FOUND;
     uint16_t offset = aMessage.GetOffset();
     uint16_t end    = aMessage.GetLength();
     Tlv      tlv;
+    uint16_t size;
 
-    while (offset + sizeof(tlv) <= end)
+    while (true)
     {
-        uint16_t length;
-
-        aMessage.Read(offset, sizeof(tlv), &tlv);
-        offset += sizeof(tlv);
-        length = tlv.GetLength();
+        VerifyOrExit(offset + sizeof(Tlv) <= end);
+        aMessage.Read(offset, sizeof(Tlv), &tlv);
 
-        if (length == kExtendedLength)
+        if (tlv.mLength != kExtendedLength)
         {
-            VerifyOrExit(offset + sizeof(length) <= end);
-            aMessage.Read(offset, sizeof(length), &length);
-            offset += sizeof(length);
-            length = HostSwap16(length);
+            size = tlv.GetSize();
+        }
+        else
+        {
+            ExtendedTlv extTlv;
+
+            VerifyOrExit(offset + sizeof(ExtendedTlv) <= end);
+            aMessage.Read(offset, sizeof(ExtendedTlv), &extTlv);
+            size = extTlv.GetSize();
         }
 
-        VerifyOrExit(length <= end - offset);
+        VerifyOrExit(offset + size <= end);
```

Comment:
This change can result in overflow errors. Thoughts on reverting to the original logic (`size <= end - offset`)?

Reply:
Thanks! Great point! will change it.